### ISSUE TITLE
CA1508 false positive in foreach loop

### ISSUE THREAD

**Issue by SebastianSchumann** [dotnet/roslyn-analyzers#3860]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

Example: [CA1508](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1508)

#### Repro steps

    Queue<Exception> exceptions = null;
    foreach (var item in disposables)
    {
        try
        {
            await item.DisposeAsync().ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            if (exceptions is null)
            {
                exceptions = new Queue<Exception>();
            }

            exceptions.Enqueue(ex);
        }
    }

The analyzer reports the `if` condition `exceptions is null` to be always true. But that is only true for the first loop that reaches this statement. The second loop will find an initialized instance for `exceptions`.

Even if you change it to `exceptions ??= new Queue<Exception>();` without that `if`-statement the error will not disappear.

**Comment by Evangelink** [comment]
That's interesting! @mavasani What's the current approach of the dataflow regarding loops? Do you explore it only once?

**Comment by mavasani** [comment]
> What's the current approach of the dataflow regarding loops?

Our DFA is based on standard dataflow algorithm works on a monotonic transfer function that relies on convergence with each iteration. We will iterate on CFG paths until we converge.

I believe the core issue here is coming from try/catch within a loop. CFG exposed by Roslyn does not have explicit structure for try/catch and we try to perform our best case DFA walking in such cases. I suspect the fix will be somewhere within the below method: https://github.com/dotnet/roslyn-analyzers/blob/303d517d74aaaade61c2e46c9d8965c23758ee7d/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowAnalysis.cs#L177-L189

**Comment by Thieum** [comment]
The same pattern triggers a false positive with strings:

```
string deviceInfo = string.Empty;
foreach (string name in propertyNames)
{
    if (propMap.ContainsKey(name))
    {
        deviceInfo = string.IsNullOrEmpty(deviceInfo) ? $"{name}: {propMap[name]}" : $"{deviceInfo}; {name}: {propMap[name]}";
    }
}
```

**Comment by mavasani** [comment]
This repros on the latest .NET8 preview SDK, but not with latest NuGet package out of main branch: https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet8/NuGet/Microsoft.CodeAnalysis.NetAnalyzers/overview/8.0.0-preview.23402.3

Fix should be part of the upcoming .NET8 preview SDK
